receiver/prometheusreceiver/internal/metrics_adjuster.go

```diff
@@ -0,0 +1,169 @@
+		logger:    logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their values based on the initial points in the
+// metricMap. If the metric is the first point in the timeseries, or the timeseries has been reset, it is
+// removed from the sequence and added to the the metricMap.
+func (ma *MetricsAdjuster) AdjustMetrics(metrics []*metricspb.Metric) []*metricspb.Metric {
+	var adjusted = make([]*metricspb.Metric, 0, len(metrics))
+	for _, metric := range metrics {
+		if ma.adjustMetric(metric) {
+			adjusted = append(adjusted, metric)
+		}
+	}
+	return adjusted
+}
+
+func (ma *MetricsAdjuster) adjustMetric(metric *metricspb.Metric) bool {
+	switch metric.MetricDescriptor.Type {
+	case metricspb.MetricDescriptor_GAUGE_DOUBLE:
+		return true
+	default:
+		return ma.adjustMetricTimeseries(metric)
+	}
+}
+
+func (ma *MetricsAdjuster) adjustMetricTimeseries(metric *metricspb.Metric) bool {
+	filtered := make([]*metricspb.TimeSeries, 0, len(metric.GetTimeseries()))
+	for _, timeseries := range metric.GetTimeseries() {
+		key := getSignature(metric.GetMetricDescriptor().GetName(), timeseries.GetLabelValues())
+		initial, ok := (*ma.metricMap)[key]
+		if ok {
+			if !ma.adjustTimeseries(metric.MetricDescriptor.Type, timeseries, initial) {
+				// detected a reset
+				(*ma.metricMap)[key] = timeseries
+			}
+			filtered = append(filtered, timeseries)
+		} else {
+			(*ma.metricMap)[key] = timeseries
+		}
+	}
+	if len(filtered) == 0 {
+		return false
+	}
+	metric.Timeseries = filtered
+	return true
+}
+
+func (ma *MetricsAdjuster) adjustTimeseries(metricType metricspb.MetricDescriptor_Type, current, initial *metricspb.TimeSeries) bool {
+	if !ma.adjustPoints(metricType, current.GetPoints(), initial.GetPoints()) {
+		return false
+	}
+	current.StartTimestamp = initial.StartTimestamp
+	return true
+}
+
+func (ma *MetricsAdjuster) adjustPoints(metricType metricspb.MetricDescriptor_Type, current, initial []*metricspb.Point) bool {
+	if len(current) != len(initial) {
+		// this shouldn't happpen, both lengths should always be 1.
+		ma.logger.Info("len(current points) != len(initial points)", len(current), len(initial))
+		return true
+	}
+	for i := 0; i < len(current); i++ {
+		if !ma.adjustPoint(metricType, current[i], initial[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// Types of metrics returned supported by prometheus:
+// - MetricDescriptor_CUMULATIVE_DOUBLE
+// - MetricDescriptor_GAUGE_DOUBLE
+// - MetricDescriptor_CUMULATIVE_DISTRIBUTION
+// - MetricDescriptor_GAUGE_DISTRIBUTION
+// - MetricDescriptor_SUMMARY
+func (ma *MetricsAdjuster) adjustPoint(metricType metricspb.MetricDescriptor_Type, current, initial *metricspb.Point) bool {
+	switch metricType {
+	case metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		// note: sum of squared deviation not currently supported
+		initialDist := initial.GetDistributionValue()
+		currentDist := current.GetDistributionValue()
+		if currentDist.Count < initialDist.Count {
```

Comment:
it should be if either count or sum is less than previous value

Reply:
Done